### PR TITLE
get CommsDebugMode to work with DTensor

### PR DESCRIPTION
**Pull request by bdhirsh**
Tested with Wanchao's repro:
```
from typing import Tuple, List, Dict, cast
import torch
import torch.nn as nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed._tensor import distribute_tensor, DTensor, Shard, Placement, Replicate

mesh = init_device_mesh(device_type="cuda", mesh_shape=(2,))
x = torch.randn(4, 8, requires_grad=True)
y = torch.randn(4, 32, requires_grad=True)
x_dtensor = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
y_dtensor = DTensor.from_local(y, mesh, [Shard(0)], run_check=False)
from torch.distributed._tensor.debug import CommDebugMode
comm_mode = CommDebugMode()
with comm_mode:
    z = torch.mm(x_dtensor, y_dtensor)
print(comm_mode.get_comm_counts())
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118769

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Review by wanchaol — COMMENTED** [review]
Thanks for fixing this!! wondering if you could add a test to `test_comm_mode.py`?

File: torch/distributed/_tensor/debug/comm_mode.py [hunk 1]
```diff
@@ -60,6 +61,12 @@ def __exit__(self, *args):
         super().__exit__(*args)
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        # When running this mode with DTensor, ordinarily all modes will
+        # run **before** subclasses get a chance to run.
+        # Returning NotImplemented here gives us a chance to let DTensor
+        # run and desugar into comms ops, before CommDebugMode sees them.
+        if any(t == DTensor for t in types):
+            return NotImplemented
```
[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/distributed/_tensor/debug/comm_mode.py:69]: Yep - the idea is that we don't actually remove `CommDebugMode` from the mode stack permanently - we temporarily pop it off the stack while executing it - but if we find that it returns `NotImplemented`, "step back out", pushing the mode back on the stack and giving subclasses a chance to run their `__torch_dispatch__` first (and then when they perform a redispatch, the `CommDebugMode` will be back on the stack)

**Comment by bdhirsh** [comment]
Just added a test (pytorch is still building locally for me, will make sure it's passing once build is done)

**Review by wanchaol — APPROVED** [review]
looks great!

**Comment by awgu** [comment]
@bdhirsh @wanchaol I think that the layer norm forward and backward are introducing an unexpected collective, which was not caught previously due to `CommDebugMode` not seeing functional collectives in `__torch_dispatch__`.

Shall we just relax the test for now (P1188558579) to land this and fix layer norm in a follow-up?

**Comment by bdhirsh** [comment]
Updated - thanks for investigating the failure @awgu !

**Pull request closed without merging**